### ISSUE TITLE
transaction validation exception

### ISSUE THREAD

**Issue by nsjames** [EOSIO/eosjs#64]
**Version: 8.0.1**

Error Message
```
code: 3030000, name: "transaction_exception", what: "transaction validation exception"
message: "condition: assertion failed: read"
```
![image](https://user-images.githubusercontent.com/10360675/39335453-d0d662b6-49bb-11e8-95f0-6bf76cbb66e6.png)

It seems like perhaps nodeos isn't liking the packing for this method.
When sending through `cleos` this works fine.

eosjs
```
// Method use
contract.user(user, user.key, sig, options)...

// Verified using `fc.fromBuffer`
{ user:
   { keyid: '0',
     key: 'EOS5YfsbxrWXPb6XnKy7R2JMV41JeLNsoePuGxvE1tenD4WEqJBL5',
     name: 'TestName',
     type: 'Whatever',
     bio: '',
     links: [],
     last_active: '0' },
  strkey: 'EOS5YfsbxrWXPb6XnKy7R2JMV41JeLNsoePuGxvE1tenD4WEqJBL5',
  sig: 'EOSKXwVayHuwGMmuFXEjdwRRjFZXe9f8TsnGnXgkmkptGP3zwWkbpFmfVMTps5M8qRFB7yvVQn5nVyyhiZwCyYZT7MdiN3bjB' }

```

cleos
```
cleos push action hackathon user '{"user":{
"keyid": 0,
"key":"EOS5YfsbxrWXPb6XnKy7R2JMV41JeLNsoePuGxvE1tenD4WEqJBL5",
"name":"TestName",
"type":"Whatever",
"bio": "",
"links": [],
"last_active": 0 
}, "strkey":"EOS5YfsbxrWXPb6XnKy7R2JMV41JeLNsoePuGxvE1tenD4WEqJBL5", 
"sig":"EOSKXwVayHuwGMmuFXEjdwRRjFZXe9f8TsnGnXgkmkptGP3zwWkbpFmfVMTps5M8qRFB7yvVQn5nVyyhiZwCyYZT7MdiN3bjB"}' -p eosio
```

Relevant ABI
```
"types": [{
    "new_type_name": "uuid",
    "type": "uint64"
}]
```
```
"name": "User",
"base": "",
"fields": [{
    "name": "keyid",
    "type": "uuid"
},{
    "name": "key",
    "type": "public_key"
},{
    "name": "name",
    "type": "string"
},{
    "name": "type",
    "type": "string"
},{
    "name": "bio",
    "type": "string"
},{
    "name": "links",
    "type": "Link[]"
},{
    "name": "last_active",
    "type": "uint64"
}]
```
```
{
"name": "user",
"base": "",
"fields": [{
    "name": "user",
    "type": "User"
},{
    "name": "strkey",
    "type": "string"
},{
    "name": "sig",
    "type": "signature"
}]
}
```

From empty action method `contract.user()`
```
EXAMPLE
{
    "user": {
        "keyid": "0",
        "key": "EOS6MRy..",
        "name": "",
        "type": "",
        "bio": "",
        "links": [
            {
                "name": "",
                "url": ""
            }
        ],
        "last_active": "0"
    },
    "strkey": "",
    "sig": "SIGnature.."
}
```

**Comment by nsjames** [comment]
This happens due to the "signature" field. When removed it works fine.

**Comment by jcalfee** [comment]
Do you know the matching eosio branch or tag?  I want to lookup wasm_interface.cpp line 808 ..

**Comment by nsjames** [comment]
I'm on master but it looks like i'm a bit behind. I'll pull and rebuild and try again. 

`Your branch is behind 'origin/master' by 273 commits, and can be fast-forwarded.`

**Comment by nsjames** [comment]
Yup looks like there's still an issue

![image](https://user-images.githubusercontent.com/10360675/39387974-7d19d408-4a85-11e8-8a54-8c26b492baf6.png)

![image](https://user-images.githubusercontent.com/10360675/39387982-89f9fcb6-4a85-11e8-929f-048a459b995c.png)

**Comment by nsjames** [comment]
Here's a comment from **Jonathan Giszczak** about this on Telegram

> I'd say the signature hasn't been packaged up quite right.  The actual error looks more or less unrelated, or misdirected, as if it were trying to interpret the signature as a contract action.

**Comment by jcalfee** [comment]
npm published eosjs@9.0.1 (eosjs@dawn3) .. should be fixed

**Comment by nsjames** [comment]
![image](https://user-images.githubusercontent.com/10360675/39389410-db1a237c-4a90-11e8-998a-38719bca4683.png)

Yup that was definitely it! Thanks for the quick turnaround.

By the way it seems `ecc` is not defined in the [structs.js](https://github.com/EOSIO/eosjs/blob/master/src/structs.js#L1) which leads to an error [when trying to use it](https://github.com/EOSIO/eosjs/blob/master/src/structs.js#L325). I've put up a pull request fix for it but you might want to pull out the deeper `Signature` object instead.
